### PR TITLE
Initial rapid power-up development guide

### PR DESCRIPTION
**Pull request by invisig0th**

### CODE REVIEW

File: docs/synapse/devguides/power-ups.rst [hunk 1]
```diff
@@ -0,0 +1,426 @@
...
+A **Storm Package** consists of a YAML file which defines the various commands, modules,
+documentation, and workflows embedded within the package. 
+
+Minimal Example
+---------------
+
+As you can see in the minimal example below, the Storm Package is defined by a YAML file
+that gets processed and loaded into your Cortex.
+
+``acme-hello.yaml``::
```
Comment (vEpiphyte) [inline docs/synapse/devguides/power-ups.rst:31]: For the file content examples we can inline them directly from disk

File: docs/synapse/devguides/power-ups.rst [hunk 2]
```diff
@@ -0,0 +1,426 @@
...
+    storm>
+
+Directory Layout
+----------------
+
+Storm Modules
+=============
+
+Deploying **Storm** modules allows you to author powerful library functions that you can use in
+automation or **Storm** commands to facilitate code re-use and enforce privilidge separation boundaries.
```
Comment (vEpiphyte) [inline docs/synapse/devguides/power-ups.rst:124]: privilidge

**Review by Cisphyx — COMMENTED** [review]
Could add some info for `onload` to the advanced section, potentially with how it is used for extended model additions

File: docs/synapse/devguides/power-ups.rst [hunk 3]
```diff
@@ -0,0 +1,426 @@
...
+    function getFooByBar(bar) {
+
+        $apikey = $lib.globals.get(acme:hello:apikey)
+        $headers = ({"apikey": $apikey})
+
+        $url = $lib.str.format("https://acme.newp/api/v1/foo/{bar}", bar=$bar)
+
+        $resp = $lib.inet.http.get($url, headers=$headers)
+        if ($resp.code != 200) {
+            $lib.warn("/api/v1/foo returned HTTP code: {code}", code=$resp.code)
```
Comment (mikemoritz) [inline docs/synapse/devguides/power-ups.rst:188]: Minor nit: probably want to return null here or show an example of try/catch to attempt to get a json error message.

File: docs/synapse/devguides/power-ups.rst [hunk 4]
```diff
@@ -0,0 +1,426 @@
...
+Command line options are available within the **Storm** command by accessing the implicit
+``$cmdopts`` variable.
+
+``storm/commands/acme.hello.omgopts.storm``::
+
+    // An init {} block only runs once even if there are multiple nodes in the pipeline.
+    init {
+
+        // set global debug (once) if the user specified --debug
+        if $cmdopts.debug { $lib.debug = $lib.true } }
```
Comment (mikemoritz) [inline docs/synapse/devguides/power-ups.rst:289]: Maybe show an example of the intended use of $lib.debug

File: docs/synapse/devguides/power-ups.rst [hunk 5]
```diff
@@ -0,0 +1,426 @@
...
+
+        if ($cmdopts.hehe) { $lib.print("User Specified hehe: {hehe}", hehe=$cmdopts.hehe) }
```
Comment (mikemoritz) [inline docs/synapse/devguides/power-ups.rst:291]: Might want to differentiate (or describe) the pattern of *not* putting certain cmdopts in init that could be set per-node

File: docs/synapse/devguides/power-ups.rst [hunk 6]
```diff
@@ -0,0 +1,426 @@
...
+  debug output in any command or functions called from the command.
+
+--yield
+  By default, a command is generally expected to yield the nodes that it recieved as input from the pipeline. In
+  some instances it is useful to instruct the command to yield the nodes it creates. For example, if you specify
+  ``inet:fqdn`` nodes as input to a DNS resolver command, it may be useful to tell the command to yield the newly
+  created ``inet:dns:a`` records rather than the input ``inet:fqdn`` nodes.  Commands frequently use the ``divert``
+  **Storm** command to implement ``--yield`` functionality.
+
+--asof <time>
```
Comment (mikemoritz) [inline docs/synapse/devguides/power-ups.rst:329]: Might be good to show or describe the convention of namespacing the cache keys

Comment (invisig0th) [reply docs/synapse/devguides/power-ups.rst:329]: I think i'll add that to advanced topics :)

File: docs/synapse/devguides/power-ups.rst (hunk 5 above)
Comment (invisig0th) [reply docs/synapse/devguides/power-ups.rst:305]: captured as an advanced topic. non-runtsafe CLI args are rough to grok :D

File: examples/power-ups/rapid/acme-hello/storm/commands/acme.hello.omgopts.storm [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+init {
+
+    // set global debug (once) if the user specified --debug
+    if $cmdopts.debug { $lib.debug = $lib.true }
+
+    if ($cmdopts.hehe) { $lib.print("User Specified hehe: {hehe}", hehe=$cmdopts.hehe) }
+
+    // Normalize the FQDN in case we want to send it to an external system
+    ($ok, $fqdn) = $lib.trycast($cmdopts.fqdn, inet:fqdn)
```
Comment (Cisphyx) [inline examples/power-ups/rapid/acme-hello/storm/commands/acme.hello.omgopts.storm:11]: trycast args are backwards

File: examples/power-ups/rapid/acme-hello/storm/commands/acme.hello.omgopts.storm [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+    if (not $ok) {
+        $lib.exit("Invalid FQDN Specified: {fqdn}", ipv4=$cmdopts.fqdn)
```
Comment (Cisphyx) [inline examples/power-ups/rapid/acme-hello/storm/commands/acme.hello.omgopts.storm:13]: wrong format arg name

**Review by Cisphyx — APPROVED** [review]

**Pull request merged**